### ISSUE TITLE
aws_vpc_ipam_pool_cidr_allocation 4.17.1 -> 4.18.0 breaking change

### ISSUE THREAD

**Issue by SizZiKe** [hashicorp/terraform-provider-aws#25268]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Affected Resource(s)

* aws_vpc_ipam_pool_cidr_allocation

### Terraform Configuration Files

Please include all Terraform configurations required to reproduce the bug. Bug reports without a functional reproduction may be closed without investigation.

Here is a snippet where we're using the resource that broke:
```hcl
resource "aws_vpc_ipam_pool_cidr_allocation" "network_services_subnets" {
  for_each       = local.network_services_subnets
  ipam_pool_id   = var.ipam_pool_id_map.subnets
  netmask_length = each.value.netmask_length
}

resource "aws_vpc_ipam_pool_cidr_allocation" "firewall_subnets" {
  depends_on     = [aws_vpc_ipam_pool_cidr_allocation.network_services_subnets]
  for_each       = local.firewall_subnets
  ipam_pool_id   = var.ipam_pool_id_map.subnets
  netmask_length = each.value.netmask_length
}

resource "aws_vpc_ipam_pool_cidr_allocation" "private_subnets" {
  depends_on     = [aws_vpc_ipam_pool_cidr_allocation.firewall_subnets]
  for_each       = local.private_subnets
  ipam_pool_id   = var.ipam_pool_id_map.subnets
  netmask_length = each.value.netmask_length
}

resource "aws_vpc_ipam_pool_cidr_allocation" "nat_subnets" {
  depends_on     = [aws_vpc_ipam_pool_cidr_allocation.private_subnets]
  for_each       = local.nat_subnets
  ipam_pool_id   = var.ipam_pool_id_map.subnets
  netmask_length = each.value.netmask_length
}

resource "aws_vpc_ipam_pool_cidr_allocation" "public_subnets" {
  depends_on     = [aws_vpc_ipam_pool_cidr_allocation.nat_subnets]
  for_each       = local.public_subnets
  ipam_pool_id   = var.ipam_pool_id_map.subnets
  netmask_length = each.value.netmask_length
}

```

### Debug Output

[4.17.1-debug.log](https://gist.github.com/SizZiKe/e09fa4ee1bc4bf2eb61c99d52f4461c7#file-4-17-1-debug-log)
[4.17.1-stdout.log](https://gist.github.com/SizZiKe/e09fa4ee1bc4bf2eb61c99d52f4461c7#file-4-17-1-stdout-log)
[4.18.0 debug log](https://gist.github.com/SizZiKe/e09fa4ee1bc4bf2eb61c99d52f4461c7#file-4-18-0-debug-log)
[4.18.0-stdout.log](https://gist.github.com/SizZiKe/e09fa4ee1bc4bf2eb61c99d52f4461c7#file-4-18-0-stdout-log)
### Expected Behavior

There should be no changes per the 4.17.1 output.

### Actual Behavior

4.18.0 caused the `aws_vpc_ipam_pool_cidr_allocation` to produce different CIDR blocks which impacts downstream resources.

### Steps to Reproduce

1. use `aws_vpc_ipam_pool_cidr_allocation` to allocate multiple CIDRs based on the netmask length with `4.17.1`
2. upgrade to `4.18.0` and `aws_vpc_ipam_pool_cidr_allocation` will change the allocations, which impacts downstream resources.

### References

* #0000

**Comment by anGie44** [comment]
Hi @SizZiKe thank you for reporting this issue and apologies you came across breaking behavior. This looks to be a consequence of https://github.com/hashicorp/terraform-provider-aws/pull/25257 ; we updated the request to directly use the `IpamPoolAllocationId` in the request instead of as a key/value pair in the request filter.

**Comment by deutmeyerbrianpfg** [comment]
Agreed!  This broke our pipelines as well.  I had to revert to a snapshot of state and downgrade the provider.   It appears terraform updated all the allocation values to be the allocation ID and CIDR of the "lowest" IP cidr pool reserved.

Here's the output from the first run on `version = 4.18.0`.  All data has been sanitized approrpriately.

```hcl
aws_vpc_ipam_pool_cidr_allocation.this["10.0.0.224/27"]: Refreshing state... [id=ipam-pool-alloc-03d96e4202473458bbcab326598940159_ipam-pool-0b1816853f420f470]
aws_vpc_ipam_pool_cidr_allocation.this["10.0.0.96/27"]: Refreshing state... [id=ipam-pool-alloc-07f227aa7972c4207954a9090f366c760_ipam-pool-0b1816853f420f470]
aws_vpc_ipam_pool_cidr_allocation.this["10.0.0.160/27"]: Refreshing state... [id=ipam-pool-alloc-0aca35969ec08420dbfd6450342255ac6_ipam-pool-0b1816853f420f470]

Note: Objects have changed outside of Terraform

Terraform detected the following changes made outside of Terraform since the
last "terraform apply":

  # aws_vpc_ipam_pool_cidr_allocation.this["10.0.0.160/27"] has changed
~ resource "aws_vpc_ipam_pool_cidr_allocation" "this" {
~cidr                    = "10.0.0.160/27" ->"10.0.0.96/27"
id                      = "ipam-pool-alloc-0aca35969ec08420dbfd6450342255ac6_ipam-pool-0b1816853f420f470"
~ipam_pool_allocation_id = "ipam-pool-alloc-0aca35969ec08420dbfd6450342255ac6" ->"ipam-pool-alloc-07f227aa7972c4207954a9090f366c760"
# (4 unchanged attributes hidden)
    }

  # aws_vpc_ipam_pool_cidr_allocation.this["10.0.0.224/27"] has changed
~ resource "aws_vpc_ipam_pool_cidr_allocation" "this" {
~cidr                    = "10.0.0.224/27" ->"10.0.0.96/27"
id                      = "ipam-pool-alloc-03d96e4202473458bbcab326598940159_ipam-pool-0b1816853f420f470"
~ipam_pool_allocation_id = "ipam-pool-alloc-03d96e4202473458bbcab326598940159" ->"ipam-pool-alloc-07f227aa7972c4207954a9090f366c760"
# (4 unchanged attributes hidden)
    }

Unless you have made equivalent changes to your configuration, or ignored the
relevant attributes using ignore_changes, the following plan may include
actions to undo or respond to these changes.
─────────────────────────────────────────────────────────────────────────────
Terraform used the selected providers to generate the following execution
plan. Resource actions are indicated with the following symbols:
-/+ destroy and then create replacement

Terraform will perform the following actions:

  # aws_vpc_ipam_pool_cidr_allocation.this["10.0.0.160/27"] must be replaced
-/+ resource "aws_vpc_ipam_pool_cidr_allocation" "this" {
~cidr                    = "10.0.0.96/27" ->"10.0.0.160/27" # forces replacement
~id                      = "ipam-pool-alloc-0aca35969ec08420dbfd6450342255ac6_ipam-pool-0b1816853f420f470" ->(known after apply)
~ipam_pool_allocation_id = "ipam-pool-alloc-07f227aa7972c4207954a9090f366c760" ->(known after apply)
+resource_id             = (known after apply)
~resource_owner          = "012345678910" ->(known after apply)
~resource_type           = "custom" ->(known after apply)
# (2 unchanged attributes hidden)
    }

  # aws_vpc_ipam_pool_cidr_allocation.this["10.5.0.224/27"] must be replaced
-/+ resource "aws_vpc_ipam_pool_cidr_allocation" "this" {
~cidr                    = "10.0.0.96/27" ->"10.0.0.224/27" # forces replacement
~id                      = "ipam-pool-alloc-03d96e4202473458bbcab326598940159_ipam-pool-0b1816853f420f470" ->(known after apply)
~ipam_pool_allocation_id = "ipam-pool-alloc-07f227aa7972c4207954a9090f366c760" ->(known after apply)
+resource_id             = (known after apply)
~resource_owner          = "012345678910" ->(known after apply)
~resource_type           = "custom" ->(known after apply)
# (2 unchanged attributes hidden)
    }
```

**Comment by anGie44** [comment]
Keeping this open for now to wait for confirmation but for anyone following this issue, there has been a fix upstream (and validated with test in #25279  so no further action is needed here in the provider.

**Comment by deutmeyerbrianpfg** [comment]
I just bumped my provider to `4.21.0` and I'd agree that this appears to be resolved.  The bump resulted in no changes to my configuration.  Whew!
